File: scripts/inventory.mjs
#!/usr/bin/env node
/**
 * inventory.mjs - report the design system that already exists.
 *
 * Run before designing anything. New work should inherit tokens, components,
 * and conventions rather than overwrite them. A skill that invents a second
 * design system inside a codebase that already has one is a liability.
 *
 * Output is deliberately compact because it goes into a context window.
 */

import { readFile, readdir, stat } from 'node:fs/promises';
import { join, extname, basename, relative, dirname } from 'node:path';

const SKIP = new Set(['node_modules', 'dist', 'build', '.next', '.git', 'coverage', 'out', '.turbo', 'vendor']);
const CODE = new Set(['.css', '.scss', '.sass', '.less', '.html', '.jsx', '.tsx', '.js', '.mjs', '.cjs', '.ts', '.mts', '.cts', '.vue', '.svelte', '.astro', '.mdx']);
const CONFIGS = ['tailwind.config.js', 'tailwind.config.ts', 'tailwind.config.mjs', 'tailwind.config.cjs',
  'theme.js', 'theme.ts', 'panda.config.ts', 'uno.config.ts'];
const CONTEXT_DOCS = ['DESIGN.md', 'PRODUCT.md', 'STRUCTURE.md', 'design.md', 'product.md'];

function usage() {
  console.log(`considered inventory - reports an existing design system

Usage:
  node inventory.mjs [path] [--json]
  node inventory.mjs [--json] [path]

Arguments:
  path      source directory or individual source file (default: .)

Options:
  --json    machine-readable output
  -h, --help
            show this help`);
}

function usageError(message) {
  console.error(`considered inventory: ${message}`);
  console.error('Run "considered inventory --help" for usage.');
  process.exit(2);
}

function parseArgs(argv) {
  let target = null;
  let json = false;
  let help = false;

  for (const token of argv) {
    if (token === '--json') {
      if (json) usageError('--json may only be used once');
      json = true;
    } else if (token === '--help' || token === '-h') {
      if (help) usageError('help may only be used once');
      help = true;
    } else if (token.startsWith('-')) {
      usageError(`unknown option "${token}"`);
    } else if (target !== null) {
      usageError(`unexpected second target "${token}"`);
    } else {
      target = token;
    }
  }

  if (help && argv.length !== 1) usageError('help cannot be combined with a target or other options');
  return { target: target || '.', json, help };
}

const options = parseArgs(process.argv.slice(2));
if (options.help) {
  usage();
  process.exit(0);
}

const warnings = new Set();
const warn = (kind, path, error) => {
  const detail = error?.message ? `: ${error.message}` : '';
  const message = `Cannot read ${kind} ${path}${detail}`;
  if (warnings.has(message)) return;
  warnings.add(message);
  console.error(`warning: ${message}`);
};

async function walk(dir, acc = []) {
  let entries;
  try {
    entries = await readdir(dir, { withFileTypes: true });
  } catch (error) {
    warn('directory', dir, error);
    return acc;
  }
  for (const entry of entries) {
    if (entry.name.startsWith('.') && entry.name !== '.storybook') continue;
    const full = join(dir, entry.name);
    if (entry.isDirectory()) {
      if (!SKIP.has(entry.name)) await walk(full, acc);
    } else {
      acc.push(full);
    }
  }
  return acc;
}

async function exists(path) {
  try {
    await stat(path);
    return true;
  } catch (error) {
    if (error.code !== 'ENOENT') warn('path', path, error);
    return false;
  }
}

const textCache = new Map();
function readText(path) {
  if (!textCache.has(path)) {
    textCache.set(path, readFile(path, 'utf8').catch(error => {
      warn('file', path, error);
      return null;
    }));
  }
  return textCache.get(path);
}

const target = options.target;
let targetStat;
try {
  targetStat = await stat(target);
} catch (error) {
  console.error(`Cannot inspect ${target}: ${error.message}`);
  process.exit(2);
}

let all;
let scanRoot;
let targetKind;
if (targetStat.isDirectory()) {
  targetKind = 'directory';
  scanRoot = target;
  all = await walk(target);
} else if (targetStat.isFile()) {
  targetKind = 'file';
  scanRoot = dirname(target);
  all = [target];
} else {
  console.error(`Cannot inventory ${target}: expected a directory or regular file.`);
  process.exit(2);
}

const code = all.filter(file => CODE.has(extname(file)));

/* ---------- tokens ---------- */

const tokens = new Map();      // name -> value
const fonts = new Set();
const spacing = new Set();
const radii = new Set();

for (const file of code) {
  const text = await readText(file);
  if (text === null) continue;
  for (const match of text.matchAll(/(--[a-zA-Z0-9-]+)\s*:\s*([^;{}]+)/g)) {
    const name = match[1], value = match[2].trim();
    if (!tokens.has(name)) tokens.set(name, value);
  }
  for (const match of text.matchAll(/font-family\s*:\s*([^;{}]+)/gi)) fonts.add(match[1].trim().replace(/["']/g, ''));
  for (const match of text.matchAll(/(?:margin|padding|gap)(?:-[a-z]+)?\s*:\s*([^;{}]+)/gi)) {
    for (const value of match[1].matchAll(/(\d+(?:\.\d+)?(?:px|rem))/g)) spacing.add(value[1]);
  }
  for (const match of text.matchAll(/border-radius\s*:\s*([^;{}]+)/gi)) radii.add(match[1].trim());
}

const group = prefixes => {
  const out = {};
  for (const [name, value] of tokens) {
    const hit = prefixes.find(prefix => name.includes(prefix));
    if (hit) (out[hit] ||= []).push(`${name}: ${value}`);
  }
  return out;
};
const colorish = group(['color', 'bg', 'surface', 'text', 'border', 'accent', 'primary', 'danger', 'success', 'warning']);

/* ---------- components ---------- */

const componentDirs = new Set();
const components = new Map();  // name -> variants
for (const file of code) {
  const rel = targetKind === 'file'
    ? join(basename(dirname(file)), basename(file))
    : relative(scanRoot, file);
  if (!/components?|ui\/|design-system|primitives/i.test(rel)) continue;
  componentDirs.add(dirname(rel) === '.' ? '.' : dirname(rel).split('\\').join('/'));
  const name = basename(file, extname(file));
  if (/^(index|types|utils|constants)$/i.test(name)) continue;
  const text = await readText(file);
  if (text === null) continue;
  const variants = new Set();
  for (const match of text.matchAll(/(?:variant|appearance|kind|intent|tone)\s*[:?]?\s*["']([a-zA-Z-]+)["']/g)) variants.add(match[1]);
  for (const match of text.matchAll(/["'](primary|secondary|tertiary|ghost|subtle|outline|destructive|danger|link)["']/g)) variants.add(match[1]);
  components.set(name, [...variants]);
}

/* ---------- context docs ---------- */

const docs = [];
const configs = [];
if (targetKind === 'directory') {
  for (const doc of CONTEXT_DOCS) if (await exists(join(target, doc))) docs.push(doc);
  for (const config of CONFIGS) if (await exists(join(target, config))) configs.push(config);
} else {
  const name = basename(target);
  if (CONTEXT_DOCS.includes(name)) docs.push(name);
  if (CONFIGS.includes(name)) configs.push(name);
}

/* ---------- report ---------- */

const cap = (arr, n) => arr.length > n ? arr.slice(0, n).concat(`... +${arr.length - n} more`) : arr;

if (options.json) {
  console.log(JSON.stringify({
    target,
    targetKind,
    files: code.length,
    tokens: Object.fromEntries(tokens),
    fonts: [...fonts],
    spacing: [...spacing],
    radii: [...radii],
    components: Object.fromEntries(components),
    configs,
    docs,
    warnings: [...warnings]
  }, null, 2));
  process.exit(0);
}

const hasSystem = tokens.size > 0 || components.size > 0 || configs.length > 0;
const scope = targetKind === 'file' ? `in ${target}` : `under ${target}`;
console.log(`\n  EXISTING SYSTEM  ${code.length} source ${code.length === 1 ? 'file' : 'files'} ${scope}\n`);

if (!hasSystem) {
  console.log('  No design system detected. You are defining one.');
  console.log('  Establish tokens before components, and components before screens.\n');
} else {
  if (configs.length) console.log(`  Config      ${configs.join(', ')}`);
  if (docs.length)    console.log(`  Context     ${docs.join(', ')}  (read these before designing)`);

  if (tokens.size) {
    console.log(`\n  Tokens      ${tokens.size} custom properties`);
    for (const [key, list] of Object.entries(colorish)) {
      console.log(`    ${key.padEnd(9)} ${cap(list, 4).join('  ')}`);
    }
  }
  if (fonts.size)   console.log(`\n  Fonts       ${cap([...fonts], 4).join(' | ')}`);
  if (spacing.size) console.log(`  Spacing     ${cap([...spacing].sort((a, b) => parseFloat(a) - parseFloat(b)), 12).join(' ')}`);
  if (radii.size)   console.log(`  Radii       ${cap([...radii], 6).join(' | ')}`);

  if (components.size) {
    console.log(`\n  Components  ${components.size} in ${[...componentDirs].slice(0, 3).join(', ')}`);
    for (const [name, variants] of [...components].slice(0, 20)) {
      console.log(`    ${name.padEnd(18)} ${variants.length ? variants.join(', ') : '(no variants detected)'}`);
    }
    if (components.size > 20) console.log(`    ... +${components.size - 20} more`);
  }

  console.log('\n  Inherit this system. Do not invent a parallel one.');
  console.log('  If the assigned direction conflicts with these tokens, the tokens win');
  console.log('  unless the user explicitly asked for a redesign.\n');
}

File: crates/considered-scan/src/inventory.rs
//! Port of `scripts/inventory.mjs` — report the design system that already exists.
//!
//! Walks a directory tree, collects code files, extracts design tokens (CSS custom
//! properties, fonts, spacing, radii), detects components and configs, and returns
//! a structured [`InventoryReport`]. The crate never prints or terminates.

use considered_core::EngineError;
use regex::Regex;
use serde::ser::{SerializeMap, SerializeStruct};
use serde::{Deserialize, Serialize, Serializer};
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};
use walkdir::WalkDir;

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const SKIP_DIRS: &[&str] = &[
    "node_modules",
    "dist",
    "build",
    ".next",
    ".git",
    "coverage",
    "out",
    ".turbo",
    "vendor",
];

const CODE_EXTENSIONS: &[&str] = &[
    ".css", ".scss", ".sass", ".less", ".html", ".jsx", ".tsx", ".js", ".mjs", ".cjs", ".ts",
    ".mts", ".cts", ".vue", ".svelte", ".astro", ".mdx",
];

const CONFIG_FILES: &[&str] = &[
    "tailwind.config.js",
    "tailwind.config.ts",
    "tailwind.config.mjs",
    "tailwind.config.cjs",
    "theme.js",
    "theme.ts",
    "panda.config.ts",
    "uno.config.ts",
];

const CONTEXT_DOCS: &[&str] = &[
    "DESIGN.md",
    "PRODUCT.md",
    "STRUCTURE.md",
    "design.md",
    "product.md",
];

const TOKEN_PREFIXES: &[&str] = &[
    "color", "bg", "surface", "text", "border", "accent", "primary", "danger", "success", "warning",
];

// ---------------------------------------------------------------------------
// Output types
// ---------------------------------------------------------------------------

/// The full inventory report returned to callers.
#[derive(Debug, Clone, Deserialize)]
pub struct InventoryReport {
    pub target: String,
    pub target_kind: String,
    pub files: usize,
    pub tokens: BTreeMap<String, String>,
    pub fonts: Vec<String>,
    pub spacing: Vec<String>,
    pub radii: Vec<String>,
    pub components: BTreeMap<String, Vec<String>>,
    pub configs: Vec<String>,
    pub docs: Vec<String>,
    pub warnings: Vec<String>,
    #[serde(skip)]
    token_order: Vec<String>,
    #[serde(skip)]
    component_order: Vec<String>,
    #[serde(skip)]
    component_dirs: Vec<String>,
}

impl InventoryReport {
    /// Token groups in the same first-seen order used by the Node renderer.
    pub fn ordered_token_groups(&self) -> Vec<(String, Vec<String>)> {
        let mut groups: Vec<(String, Vec<String>)> = Vec::new();
        for name in &self.token_order {
            let Some(value) = self.tokens.get(name) else {
                continue;
            };
            let Some(prefix) = TOKEN_PREFIXES.iter().find(|prefix| name.contains(**prefix)) else {
                continue;
            };
            if let Some((_, values)) = groups.iter_mut().find(|(key, _)| key == prefix) {
                values.push(format!("{name}: {value}"));
            } else {
                groups.push((prefix.to_string(), vec![format!("{name}: {value}")]));
            }
        }
        groups
    }

    /// Component directories in first-seen order, for compatibility rendering.
    pub fn component_dirs(&self) -> &[String] {
        &self.component_dirs
    }
}

struct OrderedMap<'a, V> {
    order: &'a [String],
    values: &'a BTreeMap<String, V>,
}

impl<V: Serialize> Serialize for OrderedMap<'_, V> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut map = serializer.serialize_map(Some(self.order.len()))?;
        for key in self.order {
            if let Some(value) = self.values.get(key) {
                map.serialize_entry(key, value)?;
            }
        }
        map.end()
    }
}

impl Serialize for InventoryReport {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut state = serializer.serialize_struct("InventoryReport", 11)?;
        state.serialize_field("target", &self.target)?;
        state.serialize_field("targetKind", &self.target_kind)?;
        state.serialize_field("files", &self.files)?;
        state.serialize_field(
            "tokens",
            &OrderedMap {
                order: &self.token_order,
                values: &self.tokens,
            },
        )?;
        state.serialize_field("fonts", &self.fonts)?;
        state.serialize_field("spacing", &self.spacing)?;
        state.serialize_field("radii", &self.radii)?;
        state.serialize_field(
            "components",
            &OrderedMap {
                order: &self.component_order,
                values: &self.components,
            },
        )?;
        state.serialize_field("configs", &self.configs)?;
        state.serialize_field("docs", &self.docs)?;
        state.serialize_field("warnings", &self.warnings)?;
        state.end()
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn is_code_file(path: &Path) -> bool {
    path.extension()
        .and_then(|e| e.to_str())
        .map(|e| {
            let ext = format!(".{e}");
            CODE_EXTENSIONS.contains(&ext.as_str())
        })
        .unwrap_or(false)
}

fn collect_files(root: &Path) -> (Vec<PathBuf>, Vec<String>) {
    let mut files = Vec::new();
    let mut warnings = Vec::new();

    for entry in WalkDir::new(root).into_iter().filter_entry(|e| {
        let name = e.file_name().to_string_lossy();
        if e.depth() > 0 && name.starts_with('.') && name != ".storybook" {
            return false;
        }
        if e.depth() > 0 && e.file_type().is_dir() && SKIP_DIRS.contains(&name.as_ref()) {
            return false;
        }
        true
    }) {
        match entry {
            Ok(e) if e.file_type().is_file() => files.push(e.into_path()),
            Err(err) => warnings.push(format!("Cannot read directory: {err}")),
            _ => {}
        }
    }

    files.sort();
    (files, warnings)
}

// ---------------------------------------------------------------------------
// Token extraction
// ---------------------------------------------------------------------------

struct TokenExtraction {
    tokens: BTreeMap<String, String>,
    token_order: Vec<String>,
    fonts: Vec<String>,
    spacing: Vec<String>,
    radii: Vec<String>,
    warnings: Vec<String>,
}

fn extract_tokens(code_files: &[PathBuf], root: &Path) -> TokenExtraction {
    let re_custom_prop = Regex::new(r"(--[a-zA-Z0-9-]+)\s*:\s*([^;{}]+)").unwrap();
    let re_font_family = Regex::new(r"(?i)font-family\s*:\s*([^;{}]+)").unwrap();
    let re_spacing = Regex::new(r"(?i)(?:margin|padding|gap)(?:-[a-z]+)?\s*:\s*([^;{}]+)").unwrap();
    let re_spacing_val = Regex::new(r"(\d+(?:\.\d+)?(?:px|rem))").unwrap();
    let re_radius = Regex::new(r"(?i)border-radius\s*:\s*([^;{}]+)").unwrap();

    let mut tokens: BTreeMap<String, String> = BTreeMap::new();
    let mut token_order = Vec::new();
    let mut fonts = Vec::new();
    let mut spacing = Vec::new();
    let mut radii = Vec::new();
    let mut warnings: Vec<String> = Vec::new();

    for file in code_files {
        let text = match fs::read_to_string(file) {
            Ok(t) => t,
            Err(e) => {
                let rel = file.strip_prefix(root).unwrap_or(file);
                warnings.push(format!("Cannot read file {}: {e}", rel.display()));
                continue;
            }
        };

        for cap in re_custom_prop.captures_iter(&text) {
            let name = cap[1].to_string();
            let value = cap[2].trim().to_string();
            if !tokens.contains_key(&name) {
                token_order.push(name.clone());
                tokens.insert(name, value);
            }
        }

        for cap in re_font_family.captures_iter(&text) {
            let val = cap[1].trim().replace(['"', '\''], "");
            if !fonts.contains(&val) {
                fonts.push(val);
            }
        }

        for cap in re_spacing.captures_iter(&text) {
            for v in re_spacing_val.captures_iter(&cap[1]) {
                let value = v[1].to_string();
                if !spacing.contains(&value) {
                    spacing.push(value);
                }
            }
        }

        for cap in re_radius.captures_iter(&text) {
            let value = cap[1].trim().to_string();
            if !radii.contains(&value) {
                radii.push(value);
            }
        }
    }

    TokenExtraction {
        tokens,
        token_order,
        fonts,
        spacing,
        radii,
        warnings,
    }
}

// ---------------------------------------------------------------------------
// Component detection
// ---------------------------------------------------------------------------

fn detect_components(
    code_files: &[PathBuf],
    root: &Path,
    target_kind: &str,
) -> (BTreeMap<String, Vec<String>>, Vec<String>, Vec<String>) {
    let re_component_path = Regex::new(r"(?i)components?|ui/|design-system|primitives").unwrap();
    let re_skip_name = Regex::new(r"(?i)^(index|types|utils|constants)$").unwrap();
    let re_variant1 =
        Regex::new(r#"(?:variant|appearance|kind|intent|tone)\s*[:?]?\s*["']([a-zA-Z-]+)["']"#)
            .unwrap();
    let re_variant2 = Regex::new(
        r#"["'](primary|secondary|tertiary|ghost|subtle|outline|destructive|danger|link)["']"#,
    )
    .unwrap();

    let mut components: BTreeMap<String, Vec<String>> = BTreeMap::new();
    let mut component_order = Vec::new();
    let mut component_dirs = Vec::new();

    for file in code_files {
        let rel = if target_kind == "file" {
            let parent = file
                .parent()
                .and_then(|p| p.file_name())
                .unwrap_or_default();
            let name = file.file_name().unwrap_or_default();
            format!("{}/{}", parent.to_string_lossy(), name.to_string_lossy())
        } else {
            let r = file.strip_prefix(root).unwrap_or(file);
            r.to_string_lossy().replace('\\', "/")
        };

        if !re_component_path.is_match(&rel) {
            continue;
        }

        let dir = if target_kind == "file" {
            ".".to_string()
        } else {
            let d = file.parent().unwrap_or(root);
            d.strip_prefix(root)
                .unwrap_or(d)
                .to_string_lossy()
                .replace('\\', "/")
        };
        let dir = if dir.is_empty() { ".".to_string() } else { dir };
        if !component_dirs.contains(&dir) {
            component_dirs.push(dir);
        }

        let stem = file
            .file_stem()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string();
        if re_skip_name.is_match(&stem) {
            continue;
        }

        let text = match fs::read_to_string(file) {
            Ok(t) => t,
            Err(_) => continue,
        };

        let mut variants = Vec::new();
        for cap in re_variant1.captures_iter(&text) {
            let value = cap[1].to_string();
            if !variants.contains(&value) {
                variants.push(value);
            }
        }
        for cap in re_variant2.captures_iter(&text) {
            let value = cap[1].to_string();
            if !variants.contains(&value) {
                variants.push(value);
            }
        }

        if !components.contains_key(&stem) {
            component_order.push(stem.clone());
        }
        components.insert(stem, variants);
    }

    (components, component_order, component_dirs)
}

// ---------------------------------------------------------------------------
// Config & doc detection
// ---------------------------------------------------------------------------

fn detect_configs_and_docs(
    root: &Path,
    target_kind: &str,
    target_file: Option<&Path>,
) -> (Vec<String>, Vec<String>) {
    let mut docs = Vec::new();
    let mut configs = Vec::new();

    if target_kind == "directory" {
        for doc in CONTEXT_DOCS {
            if root.join(doc).exists() {
                docs.push(doc.to_string());
            }
        }
        for cfg in CONFIG_FILES {
            if root.join(cfg).exists() {
                configs.push(cfg.to_string());
            }
        }
    } else if let Some(file) = target_file {
        let name = file
            .file_name()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string();
        let name_ref = name.as_str();
        if CONTEXT_DOCS.contains(&name_ref) {
            docs.push(name.clone());
        }
        if CONFIG_FILES.contains(&name_ref) {
            configs.push(name);
        }
    }

    docs.sort();
    configs.sort();
    (configs, docs)
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Run the inventory scan on `target`.
///
/// `target` may be a directory or a single file. Returns a structured
/// [`InventoryReport`] with all detected design-system artifacts.
pub fn inventory(target: &Path) -> Result<InventoryReport, EngineError> {
    let meta = fs::symlink_metadata(target)
        .map_err(|e| EngineError::Usage(format!("Cannot inspect {}: {e}", target.display())))?;
    if meta.file_type().is_symlink() {
        return Err(EngineError::Usage(format!(
            "refusing symlink target: {}",
            target.display()
        )));
    }

    let (target_kind, scan_root, code_files) = if meta.is_dir() {
        let (all, _) = collect_files(target);
        let code: Vec<PathBuf> = all.into_iter().filter(|f| is_code_file(f)).collect();
        ("directory".to_string(), target.to_path_buf(), code)
    } else if meta.is_file() {
        let code = if is_code_file(target) {
            vec![target.to_path_buf()]
        } else {
            Vec::new()
        };
        let root = target.parent().unwrap_or(Path::new(".")).to_path_buf();
        ("file".to_string(), root, code)
    } else {
        return Err(EngineError::Usage(format!(
            "Cannot inventory {}: expected a directory or regular file.",
            target.display()
        )));
    };

    let extracted = extract_tokens(&code_files, &scan_root);
    let (components, component_order, component_dirs) =
        detect_components(&code_files, &scan_root, &target_kind);
    let (configs, docs) = detect_configs_and_docs(
        target,
        &target_kind,
        if target_kind == "file" {
            Some(target)
        } else {
            None
        },
    );

    let mut warnings = extracted.warnings;
    warnings.sort();
    warnings.dedup();

    Ok(InventoryReport {
        target: target.to_string_lossy().to_string(),
        target_kind,
        files: code_files.len(),
        tokens: extracted.tokens,
        fonts: extracted.fonts,
        spacing: extracted.spacing,
        radii: extracted.radii,
        components,
        configs,
        docs,
        warnings,
        token_order: extracted.token_order,
        component_order,
        component_dirs,
    })
}

/// Group tokens by semantic prefix (color, bg, surface, text, border, etc.).
pub fn group_tokens_by_prefix(tokens: &BTreeMap<String, String>) -> BTreeMap<String, Vec<String>> {
    let mut out: BTreeMap<String, Vec<String>> = BTreeMap::new();
    for (name, value) in tokens {
        for prefix in TOKEN_PREFIXES {
            if name.contains(prefix) {
                out.entry(prefix.to_string())
                    .or_default()
                    .push(format!("{name}: {value}"));
                break;
            }
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    #[test]
    fn inventory_empty_dir_returns_zero_files() {
        let dir = tempfile::tempdir().unwrap();
        let report = inventory(dir.path()).unwrap();
        assert_eq!(report.files, 0);
        assert_eq!(report.target_kind, "directory");
    }

    #[test]
    fn inventory_detects_css_tokens() {
        let dir = tempfile::tempdir().unwrap();
        fs::write(
            dir.path().join("theme.css"),
            ":root { --color-primary: #007bff; font-family: Inter; margin: 8px; border-radius: 4px; }",
        )
        .unwrap();

        let report = inventory(dir.path()).unwrap();
        assert_eq!(report.files, 1);
        assert!(report.tokens.contains_key("--color-primary"));
        assert!(report.fonts.contains(&"Inter".to_string()));
        assert!(report.spacing.contains(&"8px".to_string()));
        assert!(report.radii.contains(&"4px".to_string()));
    }

    #[test]
    fn inventory_detects_config_files() {
        let dir = tempfile::tempdir().unwrap();
        fs::write(dir.path().join("tailwind.config.js"), "").unwrap();
        fs::write(dir.path().join("DESIGN.md"), "").unwrap();

        let report = inventory(dir.path()).unwrap();
        assert!(report.configs.contains(&"tailwind.config.js".to_string()));
        assert!(report.docs.contains(&"DESIGN.md".to_string()));
    }

    #[test]
    fn inventory_single_file() {
        let dir = tempfile::tempdir().unwrap();
        let file = dir.path().join("app.css");
        fs::write(&file, ":root { --bg-surface: #fff; }").unwrap();

        let report = inventory(&file).unwrap();
        assert_eq!(report.target_kind, "file");
        assert_eq!(report.files, 1);
        assert!(report.tokens.contains_key("--bg-surface"));
    }

    #[test]
    fn group_tokens_by_prefix_works() {
        let mut tokens = BTreeMap::new();
        tokens.insert("--color-primary".to_string(), "#007bff".to_string());
        tokens.insert("--bg-surface".to_string(), "#fff".to_string());
        tokens.insert("--spacing-unit".to_string(), "4px".to_string());

        let grouped = group_tokens_by_prefix(&tokens);
        assert!(grouped.contains_key("color"));
        assert!(grouped.contains_key("bg"));
        assert!(!grouped.contains_key("spacing"));
    }

    #[test]
    fn inventory_nonexistent_target_returns_error() {
        let result = inventory(Path::new("/nonexistent/path/xyz"));
        assert!(result.is_err());
    }

    #[cfg(unix)]
    #[test]
    fn inventory_rejects_symlink_root() {
        use std::os::unix::fs::symlink;

        let dir = tempfile::tempdir().unwrap();
        let outside = tempfile::tempdir().unwrap();
        fs::write(outside.path().join("outside.css"), ":root {}").unwrap();
        let link = dir.path().join("linked");
        symlink(outside.path(), &link).unwrap();
        assert!(inventory(&link).is_err());
    }
}
